transformer_engine/common/comm_gemm_overlap/userbuffers-host.cpp

```diff
@@ -0,0 +1,489 @@
+  free(memhndl);
+
+  comm->mem_ptr[hndl] = *gpubuff;
+  return comm->free_region++;
+}
+
+int allreduce_userbuff_inplace_gpu(const int handler, const int offset, const int elements,
+                                   const int blocksize, communicator *comm, cudaStream_t stream);
+int allreduce2_userbuff_inplace_gpu(const int maxcredit, const int handler, const int offset,
+                                    const int elements, const int blocksize, communicator *comm,
+                                    cudaStream_t stream, int op);
+int reducescatter2_userbuff_inplace_gpu(const int maxcredit, const int handler, const int offset,
+                                        const int elements, const int blocksize, communicator *comm,
+                                        cudaStream_t stream, int op);
+int allgather2_userbuff_inplace_gpu(const int maxcredit, const int handler, const int offset,
+                                    const int elements, const int blocksize, communicator *comm,
+                                    cudaStream_t stream, int op);
+void allreduce_nonsharp_inplace(const int handler, const int offset, const int elements,
+                                communicator *comm, cudaStream_t stream, int op) {
+  if (elements < 64) return;  // sorry not implemented
+  // if(comm->myrank==0) fprintf(stderr,"AR2(%d) user call launch_mode=%d\n",op,comm->launch_mode);
+  const int ar_nvsize = op == userbuffers_allreduceop_nonsharp ? comm->ar_nvsize : comm->ar2_nvsize;
+  int blocksize = elements * 2;
+  int maxcredit = 0;
+  const int num_nodes = op == userbuffers_allreduceop_nonsharp ? comm->num_nodes : comm->num2_nodes;
+  blocksize = (comm->nblocks - 1 + (comm->alignblock - 1 + elements * 2) / comm->alignblock) /
+              comm->nblocks;  // FIXME TUNING
+  blocksize *= comm->alignblock;
+  if (blocksize < comm->minblock) blocksize = comm->minblock;
+
+  maxcredit = (elements * 2 + blocksize - 1) / blocksize;
+  // if(maxcredit>4) maxcredit=4;
+  // if(maxcredit>4 && ar_nvsize==1) maxcredit=4;
+  size_t peerblock = sizeof(int) * NVTE_REG0_COMMBUFFER / maxcredit;  // max size we can fit
+  if (blocksize > peerblock * ar_nvsize) blocksize = peerblock * ar_nvsize;
+  // blocksize=elements*2;
+  int sms = allreduce2_userbuff_inplace_gpu(maxcredit, handler, offset, elements, blocksize, comm,
+                                            stream, op);
+
+  if (num_nodes > 1 && comm->launch_mode & NVTE_LAUNCH_CPU) {
+    if (!sms) return;
+    comm->fifo[comm->head].optype = op;
+    comm->fifo[comm->head].basecounter = comm->basecounter[op];
+    comm->fifo[comm->head].blocksize = blocksize;
+    comm->fifo[comm->head].maxcredit = maxcredit;
+    comm->fifo[comm->head].handler = handler;
+    comm->fifo[comm->head].offset = offset;
+    comm->fifo[comm->head].elements = elements;
+
+    int newhead = (comm->head + 1) & (NVTE_MAX_REQUESTS - 1);
+    while (newhead == comm->tail) {
+    }
+    comm->head = newhead;
```

Comment:
Can you explain what this is doing? It seems we're trying to implement a work queue in `comm->fifo`. However, as far as I can tell, `comm->head` only increases up to a max value of 1023, after which we'll always overwrite the last entry. It also seems we're doing a spin lock on `comm->tail`, which doesn't seem to ever be modified. I see this pattern in several other collective implementations, so I'd like to make sure it's correct.